### ISSUE TITLE
Design System: Audit Sticky Side Nav element

### ISSUE THREAD

**Issue by Aveline-art** [hackforla/website#2173]
### Overview
We need to review the current sticky side nave and create a standard for the component  so that when you see two pages with that navigation style, they function the same and that they are WCAG compliant and have the best industry practices.

### Action Items
- [x] Find all instances of the HTML, CSS, or (inclusive) JS code for the sticky side nav from all pages (one can be found on the [about page](https://www.hackforla.org/about/)) as of issue creation.
- [x] Research online on what the standards are in terms of HTML tags, accessibility, etc. for this component.
- [x] Do a write-up on your research as a comment on this issue.
  - [x] If the recommendation requires visual changes, bring to Sunday meeting to have designers create a new redesign issue for  design.
  - [x] If the recommendation requires a refactor or other non-visual changes to the component (and can be made before redesign), please release dependency on #2174
      - [x] Add a link to your write-up in issue #2174 

#### Where the sticky nav can currently be seen
 - [about page](https://www.hackforla.org/about/)
 - [Figma, Guides pages](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?node-id=695%3A0) - Alyssa  is the designer and developer who started the new guides page layout that includes the sticky menu

**Comment by Aveline-art** [comment]
@abuna1985 I have made this issue and #2174 for the side nav. Please give me feedback on whether this is up to your standards for the design system and your plans for its rollout.

**Comment by ExperimentsInHonesty** [comment]
@adrian-zaragoza I see that you removed your assignment.  Do you want this issue to be moved back to prioritized backlog?

**Comment by adrian-zaragoza** [comment]
Hi @ExperimentsInHonesty , Yes that is correct. This should be moved back to the prioritized backlog.

**Comment by A-BMT02** [comment]
I will try to work on this issue

**Comment by ExperimentsInHonesty** [comment]
Hi @A-BMT02 The issues you are eligible to work are defined by the last issue you worked on, and you progress through issues with increasing complexity in the following order:
Good first issue
Good second issue
Small
Medium
Large

Please do not do the issues out of order.  We are trying to teach you the team methodology through the issues themselves.  If you skips to a large issue, it placed a burden on the tech leads to explain things that  will be obvious if you do each size issue in turn.  It also ensures you understand what we expect and quality of contributions.

I have unassigned you from this issue.

**Comment by A-BMT02** [comment]
I didn't realize there was an order of complexity to follow
Thank you so much for the clarification 
I appreciate

**Comment by 1anya1** [comment]
Availability for this week: 8 hrs
My estimated ETA for completing this issue: Tuesday March 22, 2022

**Comment by 1anya1** [comment]
Progress: Still researching 
Blockers: None 
Availability: 8hrs 
ETA:  By monday Mar 28, 2022

**Comment by JessicaLucindaCheng** [comment]
@1anya1 Please provide an update that includes Progress, Blockers, Availability, and ETA. Thanks.

**Comment by SAUMILDHANKAR** [comment]
@1anya1 Please let us know if you are still working on this issue or your availability has changed. Will leave a slack message as well. Thank you.

**Comment by 1anya1** [comment]
@SAUMILDHANKAR Thank you for follow up. I am still workin on this issue. 
Progress: Still researching
Blockers: None
Availability: 8hrs
ETA: By monday April 25, 2022

**Comment by SAUMILDHANKAR** [comment]
@1anya1 Just wanted to follow up on your progress update. Please let us know in a comment below. Thank you.

**Comment by SAUMILDHANKAR** [comment]
@1anya1 Please let us know if you are still working on this issue or your availability has changed. Will leave a slack message as well. Thank you.

**Comment by SAUMILDHANKAR** [comment]
@1anya1 Planning to unassign you from this issue due to inactivity. Please let us know when your availability changes. Will send an email as well. Thanks.

**Comment by AzaniaBG** [comment]
Availability: 7 AM - 8 AM 10/8/2022; 7 AM - 8 AM 10/9/2022; 10 AM - 11 AM 10/10/2022; 1 - 4 PM 10/11/2022
ETA: 10/11/2022 by 7 PM

**Comment by AzaniaBG** [comment]
Weekly update: I'm almost finished with the write-up, but I have a question that I was not able to address at Tuesday's meeting. I will also be attending Sunday's meeting to discuss this issue, so my new ETA is 10/16/2022 by 10 AM. 
Availability: 6:30 - 7:30 am 10/12/2022; 1 - 2 pm  and 5 pm - 8 pm 10/13/2022; 6:30 - 7:30 am 10/15/2022

**Comment by AzaniaBG** [comment]
## Write Up:
<details> 
<summary>Research Findings:</summary>

* [WCAG success criteria](https://www.w3.org/WAI/WCAG21/quickref/#info-and-relationships) for determining information, structure, and relationships programmatically can be satisfied with `nav`. The `<nav>` element is used to group navigation links, making them easier to locate and skip past. If more than one `<nav>` element is used on one page, use the `aria-label` or `aria-labelledby` attribute. See [here](https://www.w3.org/WAI/WCAG21/Techniques/html/H97.html) for details. This can be used instead of an ARIA landmark, and is preferred when possible (see [MDN docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/navigation_role) for more information on this). [General Principles for ARIA Landmarks]( https://www.w3.org/WAI/ARIA/apg/example-index/landmarks/index.html) indicate HTML sectioning elements can be used.
* [WCAG success criteria](https://www.w3.org/WAI/WCAG21/quickref/#keyboard-accessible) for keyboard accessibility requires ensuring keyboard control for all functionality, so the sticky sidebar should incorporate that functionality.
* When creating a sticky sidebar/nav, a programmer should ensure:

    * it does not overlap with other page elements
    *  the focus is navigable and is not obscured 
    * it does not cover too much of the screen on high zoom or on small screens
    * it is visible with landscape orientation
    * it does not prohibit interactions like keyboard focus and navigation
    * it works with zoom

(see [this article](https://cerovac.com/a11y/2021/09/sticky-and-fixed-elements-may-cause-huge-accessibility-problems/) for details)

* [WCAG success criteria](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=141#use-of-color) for use of color requires a contrast ratio of 3:1 with surrounding text, as well as additional visual cues on focus for links where color alone is used to identify them (see [here](https://www.w3.org/WAI/WCAG21/Techniques/general/G183.html) for details). 
</details>

### Summary:
The CSS styles and JavaScript logic for the sticky sidebar addressed many of the problems that a sticky sidebar might cause. Therefore, my recommendations include keeping the dynamic styles using JavaScript. I also recommend adding features related to keyboard accessibility and navigation as well as using proper color contrast ratios for the links within the sidebar (according to [this color contrast checker](https://coolors.co/contrast-checker/fa114f-ffffff), the color contrast is not sufficient for font sizes below 18pt/24px). 

### Recommendations (based on Research Findings above): 
- [ ] Use `<nav>` element instead of `<div>` for correct semantic HTML
- [ ] If more than one `<nav>` element is used on one page, use the `aria-label` or `aria-labelledby` attribute to distinguish each group.
- [ ] Ensure sticky side bar is accessible with keyboard-only user
- [ ] Use visible focus to indicate which element has focus
- [ ] Ensure 3:1 color contrast ratio for sticky sidebar links with a font size below 18pt/24px (or increase font size)
* Keep the container that prevents overlap with other user interface components
* Keep the media queries that turn off sticky sidebar for mobile, smaller breakpoints, and zoom > 200%
* Keep the function that shows sticky sidebar when user scrolls up 
* Keep the redundant visual cue (left arrow) to indicate the user's location in relation to the main page content
- [ ] [Test the sticky](https://www.w3.org/WAI/WCAG21/Techniques/css/C34):
    <details>
    <summary>tests:</summary>

    * Display content on a device / user agent in portrait mode.
    * Change the orientation to landscape.
    * Check whether the sticky header and footer will be un-fixed depending on the existing media query settings.
    * Display content on a desktop / user agent at a starting viewport width of 1280x1024 CSS pixels.
    * Change the viewport size in width and height or use the zoom function of the browser.
    * Check whether the sticky header and footer will be un-fixed at specific sizes depending on the existing media query 
    settings.

   </details>

### Research links / Resources:
* [ARIA: navigation Role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/navigation_role)
* [Sticky and fixed elements may cause huge accessibility problems](https://cerovac.com/a11y/2021/09/sticky-and-fixed-elements-may-cause-huge-accessibility-problems/)
* [Sticky elements: functionality and accessibility testing](https://technology.blog.gov.uk/2018/05/21/sticky-elements-functionality-and-accessibility-testing/)
* [Sticky Headers: 5 ways to make them better](https://www.nngroup.com/articles/sticky-headers/)

**WCAG Standards:**
* [Using media queries to un-fix sticky headers / footers](https://www.w3.org/WAI/WCAG21/Techniques/css/C34)
* [Focus Appearance](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/#guideline-24-navigable)
* [How to Meet WCAG (Quick Reference)](https://www.w3.org/WAI/WCAG21/quickref/)
* [H97 Technique - grouping related links using the nav element](https://www.w3.org/WAI/WCAG21/Techniques/html/H97)
    * [ARIA11 Technique - using ARIA landmarks](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA11.html)
    * [Sufficient Techniques for Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships)
    * [ARIA Landmarks - General Principles](https://www.w3.org/WAI/ARIA/apg/example-index/landmarks/index.html)
* [Distinguishable Use of Color](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=141#distinguishable)
    * [Ensuring at least 3:1 ratio between text and background](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
    * [Using a contrast ratio of 3:1 and providing additional visual cues](https://www.w3.org/WAI/WCAG21/Techniques/general/G183.html)
    * [Link Contrast Checker - against background](https://webaim.org/resources/contrastchecker/)
    * [Link Contrast Checker - within text](https://webaim.org/resources/linkcontrastchecker/)
* [Keyboard Accessible - sufficient techniques](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=211#principle2) 
    * [Using CSS when component receives focus](https://www.w3.org/WAI/WCAG21/Techniques/css/C15.html) - [Focus visible](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible)
    * [Navigable Guideline 2.4](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/)
    * [Focus Not Obscured](https://www.w3.org/WAI/standards-guidelines/wcag/new-in-22/#2412-focus-not-obscured-minimum-aa)
    * [Draft - Understanding Focus Not Obscured](https://www.w3.org/WAI/WCAG22/Understanding/focus-not-obscured-minimum)
* [Headings - Exception for fixed page sections (e.g., sidebars)](https://www.w3.org/WAI/tutorials/page-structure/headings/#exception-for-fixed-page-sections)

**Comment by AzaniaBG** [comment]
Update: I added my Write Up via comment and Released #2174 by moving the issue to the "New Issue Approval" column. I'll be attending Sunday's meeting (tomorrow) to discuss re-design.

**Comment by AzaniaBG** [comment]
@blulady I spoke with @sijiapitts on the design team regarding re-design recommendations and she provided this [link to Figma](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?node-id=24263%3A284261) with 3 possible designs (current, larger font-size, same font-size, different color) for the sticky side nav. Should this link also be added to #2174?

**Comment by AzaniaBG** [comment]
Update: at the All-Team meeting Sunday, we discussed the new Figma designs Sijia created based on WCAG recommendations set forth in this issue. Bonnie approved one of the options and Sijia put [this final design](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?node-id=1803%3A13692) in a pink box. I'm not able to edit the Figma to add this issue number, but I gave it to, and I believe she will add it.

**Comment by sijiapitts** [comment]
Update: Issue number 2174 has been added to Figma next to the design frame.

**Comment by blulady** [comment]
Looks like amazing work as always @AzaniaBG. You did everything the issue asked and more. I love the test sections of the comment you left. Do you think I could possibly ask you to do one more thing and add this to the wiki? It would be a shame if we had all of this great info on how to make a sticky side nav and just left it in a comment on an issue somewhere.

**Comment by AzaniaBG** [comment]
> Looks like amazing work as always @AzaniaBG. You did everything the issue asked and more. I love the test sections of the comment you left. Do you think I could possibly ask you to do one more thing and add this to the wiki? It would be a shame if we had all of this great info on how to make a sticky side nav and just left it in a comment on an issue somewhere.

@blulady ~ thank you! yes, of course ~ i can add it to the wiki. Can you clarify which wiki i should add it to?

**Comment by blulady** [comment]
I think this would be a good place. What do you think? Between Styling with reusable components and How to read and interpret issue labels?
![Screenshot (273)](https://user-images.githubusercontent.com/67162265/197933997-f24c30d0-657b-4275-bce1-a406852ec897.jpg)

**Comment by AzaniaBG** [comment]
Add write up to Wiki ETA: 10/27/2022 by 5 PM; Availability: 10/27/2022 10 am - 11 am; 4 - 5 PM

**Comment by AzaniaBG** [comment]
Added link to write up in Wiki
<details>
   <summary>Wiki after link added </summary>
   
![WikiAfter](https://user-images.githubusercontent.com/52610573/198358634-8907cc8e-4adc-4419-afe9-9bff65f5b4be.jpg)

</details>

**Comment by blulady** [comment]
Thank you. I believe this issue is closed.
